src/MongoDB.EntityFrameworkCore/Storage/MongoDatabaseWrapper.cs

```diff
@@ -240,18 +217,20 @@ private static MongoUpdate ConvertModifiedEntryToMongoUpdate(string collectionNa
         IBsonDocumentSerializer entitySerializer,
         IUpdateEntry entry)
     {
-        var idProperty = entry.EntityType.GetIdProperty();
-        var idSerializationInfo = GetPropertySerializationInfo(entitySerializer, idProperty);
-
-        var idValue = entry.GetCurrentValue(idProperty);
-        var serializedIdValue = SerializeValue(idSerializationInfo.Serializer, idValue);
+        var primaryKey = entry.EntityType.FindPrimaryKey();
+        BsonValue idValue = ToBsonDocument(entry, entitySerializer, primaryKey.Properties);
+        if (primaryKey.Properties.Count == 1)
+        {
+            // if PK consist of single field - have to unwrap it
+            idValue = idValue["_id"];
+        }
 
-        return Builders<BsonDocument>.Filter.Eq("_id", serializedIdValue);
+        return Builders<BsonDocument>.Filter.Eq("_id", idValue);
     }
 
     private static BsonSerializationInfo GetPropertySerializationInfo(
         IBsonDocumentSerializer entitySerializer,
-        IReadOnlyProperty property)
+        IPropertyBase property)
```

Comment:
I think we want to stick with IReadOnlyProperty where we can.

Reply:
I was getting compile time errors with `IReadOnlyProperty`. I think EF Core has messed up a bit.